sdks/python/apache_beam/metrics/execution.py

```diff
@@ -110,48 +121,37 @@ def __repr__(self):
         self.key, str(self.committed), str(self.attempted))
 
 
-class _MetricsEnvironment(object):
-  """Holds the MetricsContainer for every thread and other metric information.
+class MetricsEnvironment(object):
+  """Holds the MetricsContainer for every step.
 
   This class is not meant to be instantiated, instead being used to keep
   track of global state.
   """
-  def __init__(self):
-    self.METRICS_SUPPORTED = False
-    self._METRICS_SUPPORTED_LOCK = threading.Lock()
-    self.PER_THREAD = threading.local()
-    self.set_container_stack()
-
-  def set_container_stack(self):
-    if not hasattr(self.PER_THREAD, 'container'):
-      self.PER_THREAD.container = []
-
-  def container_stack(self):
-    self.set_container_stack()
-    return self.PER_THREAD.container
-
-  def set_metrics_supported(self, supported):
-    self.set_container_stack()
-    with self._METRICS_SUPPORTED_LOCK:
-      self.METRICS_SUPPORTED = supported
-
-  def current_container(self):
-    self.set_container_stack()
-    index = len(self.PER_THREAD.container) - 1
-    if index < 0:
+  METRICS_SUPPORTED = False
+
+  @classmethod
+  def set_metrics_supported(cls, supported):
+    cls.METRICS_SUPPORTED = supported
+
+  @classmethod
+  def register_container(cls, step_name, container):
+    """Register container to a step name."""
+    current_tracker = statesampler.EXECUTION_STATE_TRACKERS.current_tracker()
```

Comment:
Where would be be registering containers at a point where we don't have a handle to the statesampler itself?